SWIPs/swip-forwarder.md

```diff
@@ -0,0 +1,181 @@
+---
+SWIP: 35
+title: Kademlia forwarding service
+author: Louis Holbrook <dev@holbrook.no> (https://holbrook.no)
+discussions-to: <URL>
+status: Draft
+type: Core
+created: 2019-12-11
+requires (*optional): <SWIP number(s)>
+replaces (*optional): <SWIP number(s)>
+---
+
+## Simple Summary
+
+Provide a service that provides the most eligible peers for receiving a given message.
+
+
+## Abstract
+
+Currently all components that need to send messages to peers make their own calculations to determine which peer to send to. In particular the retrieve, push sync and pss protocols all need this feature.
+
+This SWIP proposes to provide this feature as a standard service of the network package.
+
+## Motivation
+
+Avoid parallell implementations of the same feature.
+
+## Specification
+
+The reference implementation is specified in `golang`.
+
+```
+
+New(capabilityIndex) SessionInterface
+
+
+type SessionInterface interface {
+	Subscribe() <-ForwardPeer
+	Get(numberOfPeers int) ([]ForwardPeer, error)
+	Close()
+}
+
+// also implements context.Context
+type SessionContext struct {
+	CapabilityIndex string
+	SessionId string
+}
+
+type SessionRPCInterface interface {
```

Comment:
`interface` type name should not end with `Interface`. I'd also vote to reshuffle this name altogether in favor of something more meaningful?

Reply:
`ForwardSession` `ForwardSessionAPI` @acud ?